File: bridge/rate_limit.py
"""
In-memory rate limiter by IP for HTTP endpoints.
"""

import time
from collections import defaultdict
from typing import Tuple

# (timestamp, count) per key; we trim old windows
_window_sec = 60
_max_per_window = defaultdict(lambda: (0.0, 0))  # key -> (window_start, count)


def _key(ip: str, suffix: str) -> str:
    return f"{ip}:{suffix}"


def check_rate_limit(ip: str, suffix: str, max_requests: int) -> Tuple[bool, str]:
    """
    Returns (allowed, error_message).
    suffix: "rooms" | "join" | "getsocketurl"
    """
    now = time.time()
    k = _key(ip, suffix)
    start, count = _max_per_window[k]
    if now - start >= _window_sec:
        start = now
        count = 0
    count += 1
    _max_per_window[k] = (start, count)
    if count > max_requests:
        return False, "Too many requests. Try again later."
    return True, ""

File: workers/bridge/rate-limit.ts
const RATE_LIMIT_WINDOW_SEC = 60;

interface RateLimitEntry {
  count: number;
  windowStart: number;
}

const rateLimitMap = new Map<string, RateLimitEntry>();

export function getClientIp(request: Request): string {
  const forwarded = request.headers.get('x-forwarded-for');
  if (forwarded) return forwarded.split(',')[0].trim();
  return 'unknown';
}

export function checkRateLimit(
  ip: string,
  suffix: string,
  maxRequests: number
): [boolean, string] {
  const now = Date.now() / 1000;
  const key = `${ip}:${suffix}`;
  let entry = rateLimitMap.get(key);
  if (!entry || now - entry.windowStart >= RATE_LIMIT_WINDOW_SEC) {
    entry = { count: 0, windowStart: now };
    rateLimitMap.set(key, entry);
  }
  entry.count++;
  if (entry.count > maxRequests) {
    return [false, 'Too many requests. Try again later.'];
  }
  return [true, ''];
}

export const RATE_LIMIT_ROOMS = 10;
export const RATE_LIMIT_JOIN = 30;
export const RATE_LIMIT_REGISTER_SESSION = 30;
